packages/@aws-cdk/aws-ec2/lib/user-data.ts

```diff
@@ -276,3 +276,258 @@ class CustomUserData extends UserData {
     throw new Error('CustomUserData does not support addSignalOnExitCommand, use UserData.forLinux() or UserData.forWindows() instead.');
   }
 }
+
+/**
+ * Suggested content types, however any value is allowed.
+ */
+export type MultipartContentType = 'text/x-shellscript; charset="utf-8"' | 'text/cloud-boothook; charset="utf-8"' | string;
+
+/**
+ * Options when creating `MultipartUserDataPart`.
+ */
+export interface MultipartUserDataPartOptions {
+
+  /**
+   * `Content-Type` header of this part.
+   *
+   * For Linux shell scripts use `text/x-shellscript`
+   */
+  readonly contentType: MultipartContentType;
+
+  /**
+   * `Content-Transfer-Encoding` header specifying part encoding.
+   *
+   * @default undefined - don't add this header
+   */
+  readonly transferEncoding?: string;
+}
+
+/**
+ * Options when creating `MultipartUserDataPart`.
+ */
+export interface MultipartUserDataPartOptionsWithBody extends MultipartUserDataPartOptions {
+  /**
+   * The body of message.
+   *
+   * @default undefined - body will not be added to part
+   */
+  readonly body?: string,
+}
+
+/**
+ * Options when creating `MultipartUserDataPartWrapper`.
+ */
+export interface MultipartUserDataPartWrapperOptions {
+  /**
+   * `Content-Type` header of this part.
+   *
+   * For Linux shell scripts typically it's `text/x-shellscript`.
+   *
+   * @default 'text/x-shellscript; charset="utf-8"'
+   */
+  readonly contentType?: MultipartContentType;
+}
+
+/**
+ * Interface representing part of `MultipartUserData` user data.
+ */
+export interface IMultipart {
+  /**
+   * The body of this MIME part.
+   */
+  readonly body: string | undefined;
+
+  /**
+   * `Content-Type` header of this part.
+   */
+  readonly contentType: string;
+
+  /**
+   * `Content-Transfer-Encoding` header specifying part encoding.
+   *
+   * @default undefined - don't add this header
+   */
+  readonly transferEncoding?: string;
+}
+
+/**
+ * The base class for all classes which can be used as {@link MultipartUserData}.
+ */
+export abstract class MultipartUserDataPart implements IMultipart {
```

Comment:
I know this name is technically correct, but it's a beast :). Can we think of an alternative? How about `MultipartBodyPart`? Or just `BodyPart`? I think `MultipartBody` is clear enough and short enough maybe?

Also, is this the only class that implements `IMultipart`? We should either have an abstract class define the common interface, or an interface. Both seems unnecessary. Or is this in preparation of `MultipartUserData` itself implementing `IMultipart`?

Reply:
Yes, I'll drop `IMultipart`. `MultipartUserDataPartRaw` can handle use cases where the custom part has to be generated (i.e. URL download).